test/python/algorithms/test_vqd.py

```diff
@@ -0,0 +1,584 @@
+            result = vqd.compute_eigenvalues(operator=operator)
+            self.assertAlmostEqual(result.eigenvalues.real[0], -1.0, places=5)
+
+    def test_vqd_optimizer(self):
+        """Test running same VQD twice to re-use optimizer, then switch optimizer"""
+        vqd = VQD(
+            k=2,
+            optimizer=SLSQP(),
+            quantum_instance=QuantumInstance(BasicAer.get_backend("statevector_simulator")),
+        )
+
+        def run_check():
+            result = vqd.compute_eigenvalues(operator=self.h2_op)
+            np.testing.assert_array_almost_equal(
+                result.eigenvalues.real, self.h2_energy_excited, decimal=3
+            )
+
+        run_check()
+
+        with self.subTest("Optimizer re-use"):
+            run_check()
+
+        with self.subTest("Optimizer replace"):
+            vqd.optimizer = L_BFGS_B()
+            run_check()
+
+    @data(MatrixExpectation(), None)
+    def test_backend_change(self, user_expectation):
+        """Test that VQE works when backend changes."""
+        vqd = VQD(
+            k=1,
+            ansatz=TwoLocal(rotation_blocks=["ry", "rz"], entanglement_blocks="cz"),
+            optimizer=SLSQP(maxiter=2),
+            expectation=user_expectation,
+            quantum_instance=BasicAer.get_backend("statevector_simulator"),
+        )
+        result0 = vqd.compute_eigenvalues(operator=self.h2_op)
+        if user_expectation is not None:
+            with self.subTest("User expectation kept."):
+                self.assertEqual(vqd.expectation, user_expectation)
+
+        vqd.quantum_instance = BasicAer.get_backend("qasm_simulator")
+
+        # works also if no expectation is set, since it will be determined automatically
+        result1 = vqd.compute_eigenvalues(operator=self.h2_op)
+
+        if user_expectation is not None:
+            with self.subTest("Change backend with user expectation, it is kept."):
+                self.assertEqual(vqd.expectation, user_expectation)
+
+        with self.subTest("Check results."):
+            self.assertEqual(len(result0.optimal_point), len(result1.optimal_point))
+
+    def test_set_ansatz_to_none(self):
+        """Tests that setting the ansatz to None results in the default behavior"""
+        vqd = VQD(
+            k=1,
+            ansatz=self.ryrz_wavefunction,
+            optimizer=L_BFGS_B(),
+            quantum_instance=self.statevector_simulator,
+        )
+        vqd.ansatz = None
+        self.assertIsInstance(vqd.ansatz, RealAmplitudes)
+
+    def test_set_optimizer_to_none(self):
+        """Tests that setting the optimizer to None results in the default behavior"""
+        vqd = VQD(
+            k=1,
+            ansatz=self.ryrz_wavefunction,
+            optimizer=L_BFGS_B(),
+            quantum_instance=self.statevector_simulator,
+        )
+        vqd.optimizer = None
+        self.assertIsInstance(vqd.optimizer, SLSQP)
+
+    def test_aux_operators_list(self):
```

Comment:
I think the aux operators tests are a bit obsolete since that's already tested in the VQE, could we remove that?

Reply:
Bear in mind this computes values for excited states too for which the state is computed a little differently so this ensures the ax operators are computed properly for the excited states too.